includes/class-amp-theme-support.php

```diff
@@ -397,12 +404,52 @@ public static function finish_init() {
 				self::redirect_non_amp_url( current_user_can( 'manage_options' ) ? 302 : 301 );
 			}
 
+			if ( AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) && wp_is_mobile() && is_amp_available() ) {
```

Comment:
I'm using `wp_is_mobile()` as a placeholder for now. We could even apply our own filter to that to add more stringent checks.

Reply:
Where did the JS-based redirection go? That needs to remain in-place as the default redirection method. We'll only provide server-side redirection as an opt-in for sites that can safely implement it. There is another implication if going with the server-side route: namely, instead of using `sessionStorage` well need to use a cookie to capture whether the user has opted-out of the mobile site. This same cookie can be used both client-side and, if they enable server-side redirection, server-side as well.

Reply:
> Where did the JS-based redirection go?

I misunderstood your reply in https://github.com/ampproject/amp-wp/pull/4796#discussion_r432678563. In https://github.com/ampproject/amp-wp/pull/4796/commits/0da5f23bf551c900cbdcca8f648d48576e89b232, I've re-added the JS solution, with the ability to opt-in to server-side redirection via a filter.